lib/compiler/passes/generate-js.js

```diff
@@ -1,3 +1,4 @@
+// @ts-check
```

Comment:
I think there's a way to do that regardless.

The problem currently (if I understand correctly) is that `expectType<T>(foo)` is supposed to check that foo is exactly of type T. But there's no way to make that work (with that syntax) in pure typescript. So what typescript actually checks is that `foo extends T`. In particular `expectType<"foo"|"bar">("foo")` passes the typescript check, but fails the tsd check.

But if you're willing to change the syntax a little, and accept slightly less informative error messages, you can get the same checking in pure typescript:

```
function expectExact<T>() {
  return function<U extends T>(
    arg: U
  ): [T] extends [U] ? () => U : never {
    // @ts-expect-error Just for type checking
    return () => arg;
  };
}
```

Now, instead of `expectType<T>(foo)` you have to say `expectExact<T>()(foo)()` (note that its three calls in a row).

With that, you get the exact same error message if `foo` does not extend `T`. But if `T` does not extend `foo`, you get a rather more cryptic:
```
This expression is not callable.
  Type 'never' has no call signatures.
```
Another downside is that if you forget the final `()`, you silently lose the `T extends foo` side of the checking.

On the other hand, these errors would appear in VSCode as you type (and also from the build when you `npm run ts`), rather than only when you `npm test`.

Reply:
In fact, you can do a little better with the error messages:
```
function expectExact<T>() {
  return function<U extends T>(
    tValue: U
  ): [T] extends [U] ? () => U : ["Argument of type", U, "does not match", T] {
    // @ts-expect-error Just for type checking
    return () => tValue;
  };
}
```
Now it reports:
```
This expression is not callable.
  Type '["Argument of type", "foo" | "fuz", "does not match", "bar" | "foo" | "fuz"]' has no call signatures.
```
which isn't great, but does at least tell you what the problem is...

Reply:
I've put up a [pull request](https://github.com/peggyjs/peggy/pull/436) for discussion.